File: clients/lumify-sdk-python/lumify/pagination.py
"""
Cursor pagination helpers — every ``/v1`` list endpoint takes ``after_id`` +
``limit`` (max 100, default 25) and returns ``next_after_id`` (``None`` when
exhausted). Some responses also carry ``has_more``; others (events) omit it, so
callers must derive "done" from ``next_after_id is None`` rather than relying on
``has_more``.

List envelopes are not uniform: teams/players use ``data``, events uses
``events``, sports/seasons are non-cursor. Pass ``items_key`` to
:func:`iterate_items` when the array field is not ``data``.

These mirror the TS SDK's standalone ``paginate`` / ``iterateItems`` helpers.
Resource classes also expose ``.paginate()`` / ``.iterate()`` convenience
wrappers that fill in the right ``items_key`` for you.
"""

from __future__ import annotations

from typing import Any, Callable, Dict, Iterator, Optional

# A page fetcher: (after_id, limit) -> page dict.
FetchPage = Callable[[Optional[int], int], Dict[str, Any]]


def paginate(
    fetch_page: FetchPage, *, limit: int = 25, max_pages: int = 1000
) -> Iterator[Dict[str, Any]]:
    """Yield each page of a cursor-paginated list endpoint until
    ``next_after_id`` is ``None`` (or ``max_pages`` is hit)."""
    after_id: Optional[int] = None
    for _ in range(max_pages):
        page = fetch_page(after_id, limit)
        yield page
        nxt = page.get("next_after_id") if isinstance(page, dict) else None
        if nxt is None:
            return
        after_id = nxt


def iterate_items(
    fetch_page: FetchPage,
    *,
    items_key: str = "data",
    limit: int = 25,
    max_pages: int = 1000,
) -> Iterator[Any]:
    """Flatten a paginated list endpoint into a single iterator over its items."""
    for page in paginate(fetch_page, limit=limit, max_pages=max_pages):
        items = page.get(items_key) if isinstance(page, dict) else None
        if isinstance(items, list):
            for item in items:
                yield item

File: clients/lumify-sdk/src/pagination.ts
// Cursor pagination helper — every /v1 list endpoint takes `after_id` + `limit`
// (max 100, default 25) and returns `next_after_id` (null when exhausted). Some
// responses also carry `has_more`; others (events) omit it, so callers must
// derive "done" from `next_after_id == null` rather than relying on `has_more`.
//
// List envelopes are not uniform: teams/players use `data`, events uses
// `events`, sports/seasons are non-cursor (`sports`/`seasons`). Pass `itemsKey`
// to {@link iterateItems} when the array field is not `data`.

export interface CursorPage {
  next_after_id?: number | null;
  [key: string]: unknown;
}

export interface PaginateOptions {
  /** Page size per request (API max 100, default 25). */
  limit?: number;
  /** Safety cap on total pages fetched, to bound a runaway loop. Default 1000. */
  maxPages?: number;
}

export interface IterateOptions extends PaginateOptions {
  /**
   * Field on each page that holds the item array.
   * Default `"data"` (teams/players). Events use `"events"`.
   */
  itemsKey?: string;
}

/**
 * Async-generator over every page of a cursor-paginated list endpoint.
 *
 * @example
 * for await (const page of paginate((afterId, limit) => client.events.list({ ...filters, afterId, limit }))) {
 *   for (const event of page.events ?? []) { ... }
 * }
 */
export async function* paginate<TPage extends CursorPage>(
  fetchPage: (afterId: number | undefined, limit: number) => Promise<TPage>,
  options: PaginateOptions = {}
): AsyncGenerator<TPage, void, void> {
  const limit = options.limit ?? 25;
  const maxPages = options.maxPages ?? 1000;
  let afterId: number | undefined;

  for (let page = 0; page < maxPages; page++) {
    const result = await fetchPage(afterId, limit);
    yield result;
    if (result.next_after_id == null) return;
    afterId = result.next_after_id;
  }
}

/**
 * Flatten a paginated list endpoint into a single async iterator over its items.
 *
 * @example
 * // teams/players — default itemsKey "data"
 * for await (const team of iterateItems((afterId, limit) => client.teams.list({ afterId, limit }))) { ... }
 *
 * // events — items live under "events"
 * for await (const event of iterateItems(
 *   (afterId, limit) => client.events.list({ afterId, limit }),
 *   { itemsKey: "events" }
 * )) { ... }
 */
export async function* iterateItems<TItem, TPage extends CursorPage = CursorPage>(
  fetchPage: (afterId: number | undefined, limit: number) => Promise<TPage>,
  options: IterateOptions = {}
): AsyncGenerator<TItem, void, void> {
  const itemsKey = options.itemsKey ?? "data";
  for await (const page of paginate(fetchPage, options)) {
    const items = (page as Record<string, unknown>)[itemsKey];
    if (!Array.isArray(items)) continue;
    for (const item of items as TItem[]) {
      yield item;
    }
  }
}
